gio/src/task.rs

```diff
@@ -2,34 +2,59 @@
 
 use crate::AsyncResult;
 use crate::Cancellable;
-use crate::Task;
 use glib::object::IsA;
+use glib::object::ObjectType as ObjectType_;
+use glib::signal::connect_raw;
+use glib::signal::SignalHandlerId;
 use glib::translate::*;
+use glib::value::ValueType;
+use glib::Cast;
 use libc::c_void;
 use std::boxed::Box as Box_;
+use std::fmt;
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::mem::transmute;
 use std::ptr;
 
-impl Task {
+// Implemented manually to make it generic over the return type to ensure the API
+// is sound when the task is moved across threads.
+
+glib::wrapper! {
+    #[doc(alias = "GTask")]
+    pub struct Task<V: ValueType + Hash + Ord + Clone + Debug>(Object<ffi::GTask, ffi::GTaskClass>) @implements AsyncResult;
```

Comment:
Why do we need `Hash + Ord + Clone + Debug`? The corresponding impls added by `glib::wrapper!` don't really need those trait bounds as they just operate on the pointer itself. You can implement them  inside the macro via

```rust
impl<V> Clone for Task<V> { ... }
```

etc

Reply:
Well, without those it would not compile :)

I considered adding them from the macro, but I was not sure if that was too implicit

Reply:
That's what I'm saying. You should not need them here at all if you implement the traits manually inside the macro instead of using `#[derive]`.